File: MATLAB/scripts/StreamClassifier3.m
clearvars

load './LowStream.mat'
Stream{1} = Average;

load './MidStream.mat'
Stream{2} = Average;

load './HighStream.mat'
Stream{3} = Average;

clear Average

SimulatingFile = '20190520_B41_Stream_0003_Processed.mat';
CorrectClass = 3;

TriggerSelect = [2 8 32];

EpochRange = [-0.1 0.5];
BaseLineRange = [-0.1 0.5];

SimulatingRange = [0 10];

StandardizeEnable = 0;
PCAEnable = 0;
RetainingVariance = 99;

%% Applying CSP

temp = [1 2 3 1 2];

for Deviant=1:size(Stream,2)
    H{Deviant} = CSP(Stream{Deviant}.Data{Deviant},cat(3,Stream{temp(Deviant+1)}.Data{Deviant},Stream{temp(Deviant+2)}.Data{Deviant}));
    
%     for l=1:size(H{Deviant},1)
%         H{Deviant}(l,:) = H{Deviant}(l,:)./(norm(H{Deviant}(l,:)));
%     end
    
    %H{Deviant} = [H{Deviant}(1,:);H{Deviant}(2,:);H{Deviant}(end-1,:);H{Deviant}(end,:)];
end

clear temp;

for Attended = 1:size(Stream,2)
    for Deviant = 1:size(Stream,2)
        Stream{Attended}.CSP{Deviant} = [];
        for l=1:size(Stream{Attended}.Data{Deviant},3)
            Stream{Attended}.CSP{Deviant} = cat(3,Stream{Attended}.CSP{Deviant},H{Deviant}*Stream{Attended}.Data{Deviant}(:,:,l));
        end
    end
end

%% Making Training Data

for Deviant=1:size(Stream,2)
    TrainingData.X{Deviant} = [];
    TrainingData.Y{Deviant} = [];
    for Attended=1:size(Stream,2)
        TrainingData.X{Deviant} = vertcat(TrainingData.X{Deviant},Vectorizer(Stream{Attended}.CSP{Deviant}));
        if Deviant == Attended
            TrainingData.Y{Deviant} = [TrainingData.Y{Deviant}; ones(size(Vectorizer(Stream{Attended}.CSP{Deviant}),1),1)];
        else
            TrainingData.Y{Deviant} = [TrainingData.Y{Deviant}; zeros(size(Vectorizer(Stream{Attended}.CSP{Deviant}),1),1)];
        end
    end
end

%% Standardize
if StandardizeEnable == 1
    for l=1:size(TrainingData.X,2)
        [TrainingData.X{l},Standardize{l}.meanvec,Standardize{l}.stdvec] = Standardization(TrainingData.X{l});
    end
end

%% Applying PCA

if PCAEnable == 1
    for l=1:size(TrainingData.X,2)
        [pca{l}.U,pca{l}.S,pca{l}.V] = PCA(TrainingData.X{l});
        pca{l}.k = DetDimension(pca{l}.S,RetainingVariance);
        TrainingData.X{l} = TrainingData.X{l}*pca{l}.U(:,1:pca{l}.k);
    end
end

%% Designing LDA

for Deviant=1:size(TrainingData.X,2)
    MdlLinear{Deviant} = fitcdiscr(TrainingData.X{Deviant},TrainingData.Y{Deviant},'DiscrimType','pseudolinear');
end

%% Simulating

load(SimulatingFile);

for l=1:size(Trigger,2)
    if Trigger(l) ~= 0
        FirstTrigger = l;
        break
    end
end

Count = 0;
for l=(FirstTrigger-Fs):SimulatingRange(2)*Fs:size(Data,2)
    
    IntervalData = Devide(Data,l,SimulatingRange,Fs);
    IntervalTrigger = Devide(Trigger,l,SimulatingRange,Fs);
    
    for m=1:length(TriggerSelect)
        EpochData{m} = Epoch(IntervalData,IntervalTrigger,EpochRange,TriggerSelect(m),Fs);
        BaseLineEpoch{m} = BaseLine(Epoch(IntervalData,IntervalTrigger,BaseLineRange,TriggerSelect(m),Fs),EpochRange,Fs);
        EpochData{m} = EpochData{m} - BaseLineEpoch{m};
    end
    
    for m=1:size(MdlLinear,2)
        if EpochData{m} ~= 0
            
            FilteredEpochData{m} = [];
            for n = 1:size(EpochData{m},3)
                FilteredEpochData{m} = cat(3,FilteredEpochData{m},H{m}*EpochData{m}(:,:,n));
            end

            FeatureVector = Vectorizer(FilteredEpochData{m});

            if StandardizeEnable == 1
                FeatureVector = (FeatureVector - Standardize{m}.meanvec)./Standardize{m}.stdvec;
            end
            
            if PCAEnable == 1
                FeatureVector = FeatureVector*pca{m}.U(:,1:pca{m}.k);
            end
            
            [Res{m} S{m}] = predict(MdlLinear{m},FeatureVector);
            
        else
            Res{m} = 0;
            S{m} = 0;
        end
    end
    
    NullCheck = 0;
    for m=1:3
        if sum(S{m}) == 0
            NullCheck = 1;
        end
    end
    
    if NullCheck == 0
    
    Count = Count + 1;
    %fprintf('\n');
    %fprintf('-%dth Section-\n',Count);
    %fprintf('\n');
    for m=1:size(Res,2)
        P{m} = mean(Res{m});
        %fprintf('Class %d : %f\n',m,P{m});
    end
    %fprintf('\n');
    
    %for m=1:size(P,2)
    
    Score(Count,:) = [mean(S{1}(:,2)) mean(S{2}(:,2)) mean(S{3}(:,2))];
    PScore(Count,:) = [P{1} P{2} P{3}];
    
    [M(Count,:),I(Count,:)] = max(mean(Score));
    end
    %end
    
    %break
end

%temp = sum(PScore,2);
% for l=1:length(I)
%     if temp(l) == 0
%         I(l) = 0;
%     end
% end

FalseCount = 0;
for l=1:length(I)
    if I(l) ~= CorrectClass
        FalseCount = FalseCount + 1;
    end
end

I

fprintf('Accuracy : %f%%\n',(1-FalseCount/length(I))*100);

% for l=1:size(MdlLinear,2)
%     mean(Result{l})
% end

File: Python/scripts/StreamClassifier3.py
import pickle
import numpy as np
import bcipy as bci
from scipy import io
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.svm import SVC

Directory = "/home/simon/Documents/MATLAB/20181206_B33_Stream/"
SaveFileName = ["LowStream","MidStream","HighStream"]


with open(Directory+SaveFileName[0],'rb') as web:
        LoadData = pickle.load(web)
Stream = [LoadData["EpochData"]]

with open(Directory+SaveFileName[1],'rb') as web:
        LoadData = pickle.load(web)
Stream.append(LoadData["EpochData"])

with open(Directory+SaveFileName[2],'rb') as web:
        LoadData = pickle.load(web)
Stream.append(LoadData["EpochData"])

NumStream = len(Stream)

#
#   Stream[AttendedStreamNumber][Deviant][Ch][Time][Epoch]
#

SimulatingFile = "20181206_B33_Stream_0004_Processed.mat"
CorrectClass = 1

SelectedTrigger = np.array([2, 8, 32])
0
TimeRange = np.array([-0.1, 0.5])
BaseLineRange = np.array([-0.05, 0])

SimulatingRange = np.array([10])

temp = [0,1,2,0,1]

H_ = [0]*NumStream
H = [0]*NumStream
for Deviant in range(NumStream):
    Target = Stream[Deviant][Deviant]
    NonTarget = np.concatenate([Stream[temp[Deviant+1]][Deviant][:,:,:],Stream[temp[Deviant+2]][Deviant][:,:,:]],2)
    H_[Deviant] = bci.CSP(Target,NonTarget)
    
    H[Deviant] = np.reshape(H_[Deviant][0,:],[1,H_[Deviant].shape[1]])
    H[Deviant] = np.vstack((H[Deviant],np.reshape(H_[Deviant][-1,:],[1,H_[Deviant].shape[1]])))
    
del temp


CSPData = [0]*NumStream
for Attended in range(NumStream):
    CSPData[Attended] = [0]*NumStream
    for Deviant in range(NumStream):
        CSPData[Attended][Deviant] = np.zeros([H[Deviant].shape[0],Stream[Attended][Deviant].shape[1],Stream[Attended][Deviant].shape[2]])
        for i in range(Stream[Attended][Deviant].shape[2]):
            CSPData[Attended][Deviant][:,:,i] = np.dot(H[Deviant],Stream[Attended][Deviant][:,:,i])

#for i in range(3):
    #print(CSPData[i][2].shape)
    #print((bci.Vectorizer(CSPData[i][2])).shape)

TrainingDataX = [0]*NumStream
TrainingDataY = [0]*NumStream
for Deviant in range(NumStream):
    TrainingDataX[Deviant] = bci.Vectorizer(CSPData[0][Deviant])
    for Attended in range(1,NumStream):
        TrainingDataX[Deviant] = np.vstack((TrainingDataX[Deviant],bci.Vectorizer(CSPData[Attended][Deviant])))
        
for Deviant in range(NumStream):
    for Attended in range(NumStream):
        if Deviant == Attended:
            if Attended == 0:
                TrainingDataY[Deviant] = np.ones((bci.Vectorizer(CSPData[Attended][Deviant])).shape[0])
            else:
                TrainingDataY[Deviant] = np.append(TrainingDataY[Deviant],np.ones((bci.Vectorizer(CSPData[Attended][Deviant])).shape[0]))
        else:
            if Attended == 0:
                TrainingDataY[Deviant] = np.zeros((bci.Vectorizer(CSPData[Attended][Deviant])).shape[0])
            else:
                TrainingDataY[Deviant] = np.append(TrainingDataY[Deviant],np.zeros((bci.Vectorizer(CSPData[Attended][Deviant])).shape[0]))

#print(TrainingDataX[2].shape)
#print(TrainingDataY[2].shape)


# Machine Learining

clf = [0]*NumStream
for Deviant in range(NumStream):
    clf[Deviant] = LinearDiscriminantAnalysis()
    clf[Deviant].fit(TrainingDataX[Deviant],TrainingDataY[Deviant])
    #clf[Deviant] = SVC(probability=True)
    #clf[Deviant].fit(TrainingDataX[Deviant],TrainingDataY[Deviant])




matdata = io.loadmat(Directory + SimulatingFile, squeeze_me=True)
        
Data = matdata["Data"]
Fs = matdata["Fs"]
Trigger = matdata["Trigger"]
Label = matdata["Label"]

for i in range(Trigger.shape[0]):
    if Trigger[i] != 0:
        FirstTrigger = np.array([i])
        break
    

Margin = int((TimeRange[1]-TimeRange[0])*Fs)
EpochData = [0]*NumStream
BaseLine = [0]*NumStream
Result = []
for i in range((FirstTrigger-Fs)[0],Data.shape[1],(SimulatingRange*Fs)[0]):
    #print(i)
    
    IntervalData = Data[:,(i-Margin):((i+(SimulatingRange*Fs)[0])+Margin)]
    IntervalTrigger = np.hstack((np.hstack((np.zeros(Margin),Trigger[i:(i+(SimulatingRange*Fs)[0])])),np.zeros(Margin)))
    
    for j in range(NumStream):
        EpochData[j] = bci.Epoch(IntervalData,IntervalTrigger,TimeRange,SelectedTrigger[j],Fs)
        BaseLine[j] = bci.BaseLine(IntervalData,IntervalTrigger,BaseLineRange,SelectedTrigger[j],Fs)
        EpochData[j] = EpochData[j] - BaseLine[j]
        
    FilteredEpochData = [0]*NumStream
    Res = [0]*NumStream
    Score = [0]*NumStream
    for j in range(NumStream):
        FilteredEpochData[j] = np.zeros(EpochData[j].shape)
        FilteredEpochData[j] np.zeros([H[Deviant].shape[0],EpochData[j].shape[1],EpochData[j].shape[2]])
        for k in range(EpochData[j].shape[2]):
            FilteredEpochData[j][:,:,k] = np.dot(H[j],EpochData[j][:,:,k])
            
        FeatureVector = bci.Vectorizer(FilteredEpochData[j])
        
        if FilteredEpochData[j].shape[2] > 1:
            Res[j] = clf[j].predict_proba([FeatureVector[0,:]])
            for k in range(1,FilteredEpochData[j].shape[2]):
                Res[j] = np.vstack((Res[j],clf[j].predict_proba([FeatureVector[k,:]])))
        else:
            Res[j] = clf[j].predict_proba([FeatureVector])
            
        Score[j] = np.mean(Res[j][:,0],axis=0)
        I = np.argmax(Score,axis=0)
        Result.append(I)
        #I.append(np.argmax(Score,axis=0))
    
    #print(Score)
    #print(str(I+1))
    

Result = np.array(Result)

Result = Result+1

Accuracy = np.sum(Result == CorrectClass) / Result.shape[0]

print(Result)
print("Accuracy : " + str(Accuracy*100) + "%")












